pkg/agent/controller/networkpolicy/fqdn_test.go

```diff
@@ -584,3 +585,173 @@ func TestSyncDirtyRules(t *testing.T) {
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			expectedRequeryAfter: getDuration(5 * time.Second),
+		},
+		{
+			name: "stale IP with unexpired TTL are retained",
+			existingDNSCache: map[string]dnsMeta{
+				"fqdn-test-pod.lfx.test": {
+					responseIPs: map[string]ipWithExpiration{
+						"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(5 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithExpiration{},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+		},
+		{
+			name:             "existingDNSCache is empty, the new response matches a selector.",
+			existingDNSCache: map[string]dnsMeta{},
+			dnsResponseIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			expectedRequeryAfter: getDuration(5 * time.Second),
+			mockSelectorToRuleIDs: map[fqdnSelectorItem]sets.Set[string]{
+				selectorItem1: sets.New[string]("mockRule1"),
+			},
+		},
+		{
+			name:             "existingDNSCache is empty, the new response doesn't match any selector",
+			existingDNSCache: map[string]dnsMeta{},
+			dnsResponseIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			expectedIPs: map[string]ipWithExpiration{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+			},
+			mockSelectorToRuleIDs: map[fqdnSelectorItem]sets.Set[string]{
+				selectorItem2: sets.New[string]("mockRule2"),
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			fakeClock := clocktesting.NewFakeClock(currentTime)
+			controller := gomock.NewController(t)
+			f, _ := newMockFQDNController(t, controller, nil)
+			f.clock = fakeClock
+			f.dnsEntryCache = tc.existingDNSCache
+			if tc.mockSelectorToRuleIDs != nil {
+				f.selectorItemToRuleIDs = tc.mockSelectorToRuleIDs
+			}
+
+			f.onDNSResponse(testFQDN, tc.dnsResponseIPs, nil)
+
+			cachedDnsMetaData, _ := f.dnsEntryCache[testFQDN]
+
+			if _, exists := f.selectorItemToRuleIDs[selectorItem2]; !exists {
+				assert.Equal(t, tc.expectedIPs, cachedDnsMetaData.responseIPs, "Expected %+v in cache, got %+v", tc.expectedIPs, cachedDnsMetaData.responseIPs)
+			} else {
+				assert.NotEqual(t, tc.expectedIPs, cachedDnsMetaData.responseIPs, "Did not Expect %+v in cache, got %+v", tc.expectedIPs, cachedDnsMetaData.responseIPs)
+			}
```

Comment:
It should use test cases to set expectation, instead of checking test case's parameters to set test case specific expectation.

I think the expectation is as simple as `assert.Equal(t, tc.expectedIPs, cachedDnsMetaData.responseIPs)` after you fix the `expectedIPs` in "existingDNSCache is empty, the new response doesn't match any selector" to an empty map.